### PR TITLE
Combined RepositoryFragment and RepositoryFolderFragment into a single fragment that uses compose

### PR DESCRIPTION
**Pull request by ianslinker-br**
So, in putting together the TODO list for the compose demo, I noticed that the RepositoryFragment and RepositoryFolderFragment were virtually the same UI with only a minor change to the call being made and where the header was being pulled from. Since I was going to be gutting the UI anyway, this seemed like a good opportunity to combine the code into one fragment that could handle both scenarios.

The new fragment, RepositoryBrowserFragment, is driven by a data class that contains the selected repo name and optional params for a particular folder. If the folder params are missing, it makes the base level repo call and behaves as RepositoryFragment did previously. If the folder params are non-null, then it performs the folder call and behaves as RepositoryFolderFragment did. While coding, I noticed RepositoryFolderFragment was not able to do more than one level deep due since it did not handle click events. RepositoryBrowserFragment solves for this with a recursive action back to itself if a folder is clicked. 

To prevent any sort of confusion with the old fragments, I removed them from the nav graph and removed their classes and layouts.

As far as the actual compose work goes, there was one hurdle that had to be addressed: an extra color. Up til now I have been able to shoehorn the colors needed into material theme, but this tertiary color really didn't fit any of the existing properties of a colors object. To solve for this I created an ArchitectureDemoColors class that has both a reference to the tertiary color as well as the material theme colors object. The class has properties for easy access to material color properties and there is a local provider and helper property to make it accessible where needed.

Another minor hurdle was the lack of a compose function for getting a plural resource. I have added a resource utils file with a pluralsResource function that will do this until Google adds one and structured it in the same way that the stringResource function works under thee hood.

### CODE REVIEW

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryBrowserFragment.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View {
+        val data = args.data
+
+        activityViewModel.setTitle(data.folderPath ?: data.repoName)
+        activityViewModel.showToolbar(true)
+        fragmentViewModel.getFiles(data)
+
```
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryBrowserFragment.kt:33]: Check out ComposeScreen function.   Used in AuthCodeFragment. Handles much of the boiler plate here

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryBrowserViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                } else {
+                    repo.getSource(slug, name)
+                }
+                when (result) {
+                    is ApiResult.Success -> srcFiles.set(result.data)
+                    is ApiResult.Failure -> {
+                        // TODO: Improve error messaging
+                        withContext(dispatcherProvider.Main) {
+                            toaster.toast(R.string.error_loading_repository)
+                        }
```
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryBrowserViewModel.kt:57]: I'll be doing some live coding today during Kotlin training this afternoon and creating some standard error handling for this pattern.

File: compose/src/main/java/com/bottlerocketstudios/compose/resources/Colors.kt [hunk 3]
```diff
@@ -15,14 +16,61 @@ val tertiary = Color(0xff1fada8)
...
+        background = background_black,
+        surface = white,
+        onPrimary = white,
+        onSecondary = white,
+        onBackground = black,
+        onSurface = greyish_brown,
+    )
 )
+
+data class ArchitectureDemoColors(
```
Comment (BottleRocket-Colin) [inline compose/src/main/java/com/bottlerocketstudios/compose/resources/Colors.kt:34]: Nice extension to theme colors.

**Review by BottleRocket-Colin — APPROVED** [review]

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryBrowserViewModel.kt (hunk 2 above)
Comment (ianslinker-br) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryBrowserViewModel.kt:57]: Unfortunately I had to drop early from the training to get to a client meeting so I missed the payoff on this. :(

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryBrowserFragment.kt (hunk 1 above)
Comment (ianslinker-br) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryBrowserFragment.kt:33]: Ah good call. I updated to use lauchIO but forgot about ComposeScree. Updating now.

File: app/src/main/res/navigation/nav_graph.xml [hunk 4]
```diff
@@ -35,27 +35,15 @@
...
-        android:name="com.bottlerocketstudios.brarchitecture.ui.repository.RepositoryFragment"
-        android:label="@string/empty_label"
-        tools:layout="@layout/repository_fragment">
         <action
-            android:id="@+id/action_repositoryFragment_to_repositoryFileFragment"
-            app:destination="@id/repositoryFileFragment" />
-        <action
-            android:id="@+id/action_repositoryFragment_to_repositoryFolderFragment"
-            app:destination="@id/repositoryFolderFragment" />
+            android:id="@+id/actionnHomeToRepositoryBrowserFragment"
```
Comment (jaredstockton) [inline app/src/main/res/navigation/nav_graph.xml]: @ianslinker-br nit: Refactor to `actionHomeToRepositoryBrowserFragment` (including usage in `HomeFragment`)

File: app/src/main/res/navigation/nav_graph.xml [hunk 5]
```diff
@@ -115,4 +94,18 @@
...
+    <fragment
+        android:id="@+id/repositoryBrowserFragment"
+        android:name="com.bottlerocketstudios.brarchitecture.ui.repository.RepositoryBrowserFragment"
+        android:label="@string/empty_label">
+        <argument
+            android:name="data"
+            app:argType="com.bottlerocketstudios.brarchitecture.ui.repository.RepositoryBrowserData" />
+        <action
+            android:id="@+id/action_repositoryBrowserFragment_self"
+            app:destination="@id/repositoryBrowserFragment" />
```
Comment (jaredstockton) [inline app/src/main/res/navigation/nav_graph.xml:106]: @ianslinker-br Interesting use case. Nice!

File: compose/src/main/java/com/bottlerocketstudios/compose/util/ResourceUtils.kt [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package com.bottlerocketstudios.compose.util
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/util/ResourceUtils.kt:1]: Nice addition! Interesting that Compose doesn't already have this.

File: compose/src/main/java/com/bottlerocketstudios/compose/widgets/Text.kt [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import androidx.compose.material.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.text.TextStyle
+
+@Composable
+fun IconText(
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/widgets/Text.kt:24]: @ianslinker-br @BottleRocket-Colin How do you feel about having previews for building block Composables like this one (likely in the same file)?

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragment.kt [hunk 8]
```diff
@@ -23,16 +19,6 @@ class SnippetsFragment : BaseDataBindingFragment<SnippetsFragmentViewModel, Snip
         binding.apply {
             snippetList.adapter = GroupAdapter<GroupieViewHolder>().apply {
                 add(fragmentViewModel.snippetGroup)
-                setOnItemClickListener { item, _ ->
```
Comment (jaredstockton) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragment.kt:26]: @ianslinker-br Was this code doing anything before it was removed? If so, please add a TODO to add click handling back in the future.

**Review by jaredstockton — CHANGES_REQUESTED** [review]
Appreciate the consolidation of two very similar screens into one! Left a few comments to discuss or address prior to approval :)

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragment.kt (hunk 8 above)
Comment (ianslinker-br) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragment.kt:26]: This code was checking if the view models were of type RepoFileViewModel when the view model only ever generated SnippetViewModels for the adapter. My guess is, this screen was copied from RepositoryFragment which had an identical block of code but was never updated for click code specific to snippets. As such I am not sure where, if anywhere, a snippet should navigate to when clicked and thus I removed the code.

Comment (jaredstockton) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragment.kt:26]: That definitely makes sense. Thanks for the explanation!

File: compose/src/main/java/com/bottlerocketstudios/compose/widgets/Text.kt (hunk 7 above)
Comment (ianslinker-br) [reply compose/src/main/java/com/bottlerocketstudios/compose/widgets/Text.kt:24]: Ah whoops. I had been adding them for thee other widgets I pulled out into separate files but forgot to add one here. Will add one now.

**Review by jaredstockton — APPROVED** [review]
Thanks for making the changes! Ready to go now :)

**Review by VanyaHuamanBottleRocket — APPROVED** [review]

**Pull request merged**